test/export/test_funtionalized_assertions.py

```diff
@@ -112,6 +120,88 @@ def f(x):
             "torch.ops.aten._assert_async.msg", 0, exactly=True
         ).run(gm.code)
 
+    def test_functionalization(self) -> None:
+        def f(x, y):
+            a = x.item()
+            constrain_as_size(a, 4, 7)
+            return x + 4, x + y * 2
+
+        inps = (torch.tensor([5]), torch.zeros((3, 4)))
+        ep = torch._export.export(
+            f,
+            inps,
+            constraints=[dynamic_dim(inps[1], 1) < 6],
+            _functionalize_runtime_assertions=True,
+        )
+        inps = (torch.tensor([7]), torch.ones((3, 5)))
+        self.assertTrue(torch._dynamo.utils.same(ep(*inps), (f(*inps), _DEP_TOKEN)))
+
+    def test_functionalization_with_mutated_buffer(self) -> None:
+        buf = torch.ones(6, 2)
+        weight = 0.01
+        bias = 0.2
+        d_in = 3
+        d_out = 4
+
+        class Foo(torch.nn.Module):
+            def __init__(self) -> None:
+                super().__init__()
+                self.register_buffer("buf", buf)
+
+                self.linear = torch.nn.Linear(d_in, d_out)
+                self.linear.weight.data.fill_(weight)
+                self.linear.bias.data.fill_(bias)
+
+            def forward(self, x):
+                self.buf.add_(5)
+                return self.linear(x).cos() + self.buf.sum()
```

Comment:
Can you add a test where you add the python assertion like `assert x.shape[0] > 4` or sth in the code and see it also picks up the dep token?

Reply:
Added the test case `test_native_python_assertion`. However, I still get issue (maybe it's expected though) as mentioned last week:
* If I don't specify input constraints of export, the assertion won't appear in the exported graph (instead I got `_functional_assert_async.msg` like `arg0_1.shape[0] is specialized at 2...`, which is expected)
* If I manually specify dynamic constrains like `constraints=[dynamic_dim(inp, 0), dynamic_dim(inp, 1)]`, it will actually fail me with error message `1. Could not validate (strict) constraint 2 <= L['x'].size()[0] <= oo as we generated a guard on this size variable: L['x'].size()[0] > 3.` as shown in screenshot.
And suggest me to specify input constraints as `[3 < dynamic_dim(x, 0),  dynamic_dim(x, 1)]`. This became the same as `assert x.shape[0] > 3` and it passed after I change input constraints as suggested.
![Screenshot 2023-06-20 at 10 50 37 PM](https://github.com/pytorch/pytorch/assets/4644550/53f84914-308f-4498-8650-311013b7c385)

But anyway,  I still add the test case here but in general I would say it's not really testing a new case since:
* It doesn't really covert `assert x.shape[0] > 3` to something like a inline constraint like `constrain_as_value(x.shape[0], min=4)`.

* Instead it tries to:
  * Use specified shape assertions based on example inputs during export.
  * Enforce user to set input constraints which is equivalent to native python assertion.

Reply:
It turns out this is caused by pytest assertion rewrite. Fixed by adding `PYTEST_DONT_REWRITE ...` comment.